src/util/minmax_heap.c

```diff
@@ -84,23 +80,47 @@ static void bubbleup(heap_t* h, int i) {
   }
 }
 
-int index_max_child_grandchild(heap_t* h, int i) {
+static inline int highest_index(heap_t* h, int i) {
   int a = first_child(i);
   int b = second_child(i);
-  int d = second_child(a);
   int c = first_child(a);
-  int f = second_child(b);
+  int d = second_child(a);
   int e = first_child(b);
+  int f = second_child(b);
 
-  int min_idx = -1;
-  if (a <= h->count) min_idx = a;
-  if (b <= h->count && heap_gt(h, b, min_idx)) min_idx = b;
-  if (c <= h->count && heap_gt(h, c, min_idx)) min_idx = c;
-  if (d <= h->count && heap_gt(h, d, min_idx)) min_idx = d;
-  if (e <= h->count && heap_gt(h, e, min_idx)) min_idx = e;
-  if (f <= h->count && heap_gt(h, f, min_idx)) min_idx = f;
+  if (f <= h->count) return 5;
+  if (e <= h->count) return 4;
+  if (d <= h->count) return 3;
+  if (c <= h->count) return 2;
+  if (b <= h->count) return 1;
+  if (a <= h->count) return 0;
+  return -1;
+}
 
-  return min_idx;
+int index_max_child_grandchild(heap_t* h, int i) {
+  int a = first_child(i);
+  int b = second_child(i);
+  int c = first_child(a);
+  int d = second_child(a);
+  int e = first_child(b);
+  int f = second_child(b);
+
+  switch (highest_index(h, i)) {
```

Comment:
Code duplication.

Reply:
the calculation of all the children and grandchildren is done once, as you can see [here](https://godbolt.org/z/75c818ae7)